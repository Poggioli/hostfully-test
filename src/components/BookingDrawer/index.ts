export * from './BookingDrawer'
export * from './useBookingDrawer'