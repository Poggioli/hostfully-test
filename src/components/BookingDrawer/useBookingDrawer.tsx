import { useStore } from "@/lib/store";
import { Room } from "@/service/Room";
import { useMemo } from "react";
import { useSearchParams } from "react-router-dom";

export const useBookingDrawer = () => {
  const { rooms } = useStore();
  const [searchParams, setSearchParams] = useSearchParams();
  const key = 'openBookingModal' as const

  const onClose = (): void => {
    const newSearchParams = new URLSearchParams(searchParams);
    newSearchParams.delete(key);
    setSearchParams(newSearchParams);
  }

  const onOpen = (id: string): void => {
    const newSearchParams = new URLSearchParams(searchParams)
    newSearchParams.set(key, id)
    setSearchParams(newSearchParams);
  }

  const room: Room | undefined = useMemo(() => {
    const id = searchParams.get(key) || '';
    return rooms.find((room) => room.id === id);
  }, [searchParams, rooms])

  const openBookingModal: boolean = useMemo(() => {
    return !!searchParams.get(key)
  }, [searchParams])

  return { onClose, onOpen, room, openBookingModal };
}