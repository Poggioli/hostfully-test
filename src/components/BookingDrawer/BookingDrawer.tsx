import { BookingForm } from "@/components/BookingForm";
import { Button } from "@/components/ui/button";
import { Drawer, DrawerContent, DrawerDescription, DrawerFooter, DrawerHeader, DrawerTitle } from "@/components/ui/drawer";
import { Separator } from "@/components/ui/separator";
import { FC } from "react";
import { useBookingDrawer } from "./useBookingDrawer";
import { AspectRatio } from "@/components/ui/aspect-ratio";


type BookingDrawerErrorProps = {
  onClose: () => void;
}

const BookingDrawerError: FC<BookingDrawerErrorProps> = ({ onClose }) => {
  return (
    <div>
      <DrawerHeader>
        <DrawerTitle className="text-center">Room not found</DrawerTitle>
        <DrawerDescription className="text-center">We could not found the room that you are looking for.</DrawerDescription>
      </DrawerHeader>
      <DrawerFooter>
        <Button onClick={onClose} variant="outline">Close</Button>
      </DrawerFooter>
    </div>
  )
}

const BookingDrawer: FC = () => {
  const { onClose, room, openBookingModal, booking } = useBookingDrawer();

  const onCloseInOnOpenChange = (value: boolean) => {
    if (!value) {
      onClose()
    }
  }

  return (
    <Drawer open={!!openBookingModal} onOpenChange={onCloseInOnOpenChange}>
      <DrawerContent>
        <div className="mx-auto max-w-[400px] max-h-[100vh] overflow-y-auto md:w-full md:max-w-[640px] md:overflow-hidden md:flex md:pt-8 md:items-top md:justify-center">
          {!room ? <BookingDrawerError onClose={onClose} /> : (
            <>
              <DrawerHeader className="md:max-w-[300px] md:w-full md:h-fit">
                <AspectRatio ratio={16 / 9} className="overflow-hidden mb-4 rounded-t-md">
                  <img className="rounded-md object-cover w-full h-full" src={room.photos[0]} loading="lazy" alt={`Photo of ${room.name}`} />
                </AspectRatio>
                <div className="hidden md:block">
                  <AspectRatio ratio={16 / 9} className="overflow-hidden mb-4 rounded-t-md">
                    <img className="rounded-md object-cover w-full h-full" src={room.photos[1]} loading="lazy" alt={`Photo of ${room.name}`} />
                  </AspectRatio>
                </div>
                <DrawerTitle className="text-left">{room.name}</DrawerTitle>
                <DrawerDescription className={"text-left"}>{room.description}</DrawerDescription>
              </DrawerHeader>
              <Separator className="max-w-[200px] mx-auto md:hidden" />
              <div className="p-4 md:w-full">
                <span className="text-sm mb-4 inline-block">
                  <span className="text-lg font-semibold">
                    ${room.pricePerDay}
                  </span>
                  {" "}night
                </span>
                <BookingForm room={room} booking={booking} />
              </div>
            </>
          )}
        </div>
      </DrawerContent>
    </Drawer>
  )
}

export { BookingDrawer };
