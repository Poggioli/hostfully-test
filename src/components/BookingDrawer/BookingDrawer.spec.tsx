import { act, fireEvent, render } from "@testing-library/react";
import { BookingDrawer } from "./BookingDrawer";

const mockOnOpen = jest.fn();
const mockOnClose = jest.fn();
const mockRoom = jest.fn();
const mockOpenBookingModal = jest.fn();
const mockBooking = jest.fn();
jest.mock("./useBookingDrawer", () => ({
  useBookingDrawer: () => ({
    onOpen: mockOnOpen,
    onClose: mockOnClose,
    room: (() => mockRoom())(),
    openBookingModal: (() => mockOpenBookingModal())(),
    booking: (() => mockBooking())()
  })
}));
let mockSearchParam = '';
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: () => {

    let params = new URLSearchParams(mockSearchParam)

    const setParams = (value: URLSearchParams) => {
      params = value
      mockSearchParam = value.toString();
    }

    return [
      params,
      setParams
    ];
  }
}));

const mockPostMutate = jest.fn();
const mockPutMutate = jest.fn();
jest.mock("@/service/Booking", () => ({
  usePostBooking: () => ({ mutate: mockPostMutate }),
  useEditBooking: () => ({ mutate: mockPutMutate }),
}));

describe('BookingDrawer', () => {
  it(`GIVEN a Drawer open
      WHEN room is undefined
      THEN should show error drawer`, () => {
    mockRoom.mockReturnValue(undefined);
    mockOpenBookingModal.mockReturnValue(true);
    const { baseElement } = render(<BookingDrawer />);

    expect(baseElement).toMatchSnapshot();
  });

  it(`GIVEN a Drawer open
      AND is error drawer
      WHEN user click in 'Close' button
      THEN should call onClose from useBookingDrawer`, () => {
    mockRoom.mockReturnValue(undefined);
    mockOpenBookingModal.mockReturnValue(true);
    const { getByText } = render(<BookingDrawer />);

    const cancelButton = getByText('Close');
    act(() => fireEvent.click(cancelButton));

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it(`GIVEN a Drawer open
      WHEN room is valid
      THEN should show the room infos`, () => {
    mockRoom.mockReturnValue({
      id: "id-room-1",
      description: "description room 1",
      name: "name room 1",
      photos: [],
      pricePerDay: 500,
      pricePerGuest: 100,
    });
    mockOpenBookingModal.mockReturnValue(true);
    const { baseElement } = render(<BookingDrawer />);

    expect(baseElement).toMatchSnapshot();

  })
})