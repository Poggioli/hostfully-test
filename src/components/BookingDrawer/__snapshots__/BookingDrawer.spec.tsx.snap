// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingDrawer GIVEN a Drawer open
      WHEN room is undefined
      THEN should show error drawer 1`] = `
<body
  style="position: fixed !important; top: 0px; left: 0px; right: 0px; height: auto; pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />
  <div
    aria-hidden="true"
    class="fixed inset-0 z-50 bg-black/80"
    data-aria-hidden="true"
    data-state="open"
    style="pointer-events: auto;"
    vaul-drawer-visible="true"
    vaul-overlay=""
    vaul-snap-points="false"
    vaul-snap-points-overlay="true"
  />
  <div
    aria-describedby="radix-:r2:"
    aria-labelledby="radix-:r1:"
    class="fixed inset-x-0 bottom-0 z-50 mt-24 flex h-auto flex-col rounded-t-[10px] border bg-background"
    data-state="open"
    id="radix-:r0:"
    role="dialog"
    style="pointer-events: auto;"
    tabindex="-1"
    vaul-drawer=""
    vaul-drawer-visible="true"
  >
    <div
      class="mx-auto mt-4 h-2 w-[100px] rounded-full bg-muted"
    />
    <div
      class="mx-auto max-w-[400px] max-h-[100vh] overflow-y-auto md:w-full md:max-w-[640px] md:overflow-hidden md:flex md:pt-8 md:items-top md:justify-center"
    >
      <div>
        <div
          class="grid gap-1.5 p-4 text-center sm:text-left"
        >
          <h2
            class="text-lg font-semibold leading-none tracking-tight text-center"
            id="radix-:r1:"
          >
            Room not found
          </h2>
          <p
            class="text-sm text-muted-foreground text-center"
            id="radix-:r2:"
          >
            We could not found the room that you are looking for.
          </p>
        </div>
        <div
          class="mt-auto flex flex-col gap-2 p-4"
        >
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;

exports[`BookingDrawer GIVEN a Drawer open
      WHEN room is valid
      THEN should show the room infos 1`] = `
<body
  style="right: 0px; position: fixed !important; top: 0px; left: 0px; height: auto; pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  />
  <div
    aria-hidden="true"
    class="fixed inset-0 z-50 bg-black/80"
    data-aria-hidden="true"
    data-state="open"
    style="pointer-events: auto;"
    vaul-drawer-visible="true"
    vaul-overlay=""
    vaul-snap-points="false"
    vaul-snap-points-overlay="true"
  />
  <div
    aria-describedby="radix-:r8:"
    aria-labelledby="radix-:r7:"
    class="fixed inset-x-0 bottom-0 z-50 mt-24 flex h-auto flex-col rounded-t-[10px] border bg-background"
    data-state="open"
    id="radix-:r6:"
    role="dialog"
    style="pointer-events: auto;"
    tabindex="-1"
    vaul-drawer=""
    vaul-drawer-visible="true"
  >
    <div
      class="mx-auto mt-4 h-2 w-[100px] rounded-full bg-muted"
    />
    <div
      class="mx-auto max-w-[400px] max-h-[100vh] overflow-y-auto md:w-full md:max-w-[640px] md:overflow-hidden md:flex md:pt-8 md:items-top md:justify-center"
    >
      <div
        class="grid gap-1.5 p-4 text-center sm:text-left md:max-w-[300px] md:w-full md:h-fit"
      >
        <div
          data-radix-aspect-ratio-wrapper=""
          style="position: relative; width: 100%; padding-bottom: 56.25%;"
        >
          <div
            class="overflow-hidden mb-4 rounded-t-md"
            style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
          >
            <img
              alt="Photo of name room 1"
              class="rounded-md object-cover w-full h-full"
              loading="lazy"
            />
          </div>
        </div>
        <div
          class="hidden md:block"
        >
          <div
            data-radix-aspect-ratio-wrapper=""
            style="position: relative; width: 100%; padding-bottom: 56.25%;"
          >
            <div
              class="overflow-hidden mb-4 rounded-t-md"
              style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
            >
              <img
                alt="Photo of name room 1"
                class="rounded-md object-cover w-full h-full"
                loading="lazy"
              />
            </div>
          </div>
        </div>
        <h2
          class="text-lg font-semibold leading-none tracking-tight text-left"
          id="radix-:r7:"
        >
          name room 1
        </h2>
        <p
          class="text-sm text-muted-foreground text-left"
          id="radix-:r8:"
        >
          description room 1
        </p>
      </div>
      <div
        class="shrink-0 bg-border h-[1px] w-full max-w-[200px] mx-auto md:hidden"
        data-orientation="horizontal"
        role="none"
      />
      <div
        class="p-4 md:w-full"
      >
        <span
          class="text-sm mb-4 inline-block"
        >
          <span
            class="text-lg font-semibold"
          >
            $
            500
          </span>
           
          night
        </span>
        <form
          class="flex flex-col gap-4"
        >
          <div
            class="space-y-2 flex flex-col"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
              for=":r9:-form-item"
            >
              Check-in
            </label>
            <button
              aria-controls="radix-:ra:"
              aria-describedby=":r9:-form-item-description"
              aria-expanded="false"
              aria-haspopup="dialog"
              aria-invalid="false"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[240px] pl-3 text-left font-normal text-muted-foreground"
              data-state="closed"
              id=":r9:-form-item"
              type="button"
            >
              <span>
                Pick a date to your check-in
              </span>
              <svg
                class="lucide lucide-calendar ml-auto h-4 w-4 opacity-50"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="4"
                />
                <line
                  x1="16"
                  x2="16"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="8"
                  x2="8"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="3"
                  x2="21"
                  y1="10"
                  y2="10"
                />
              </svg>
            </button>
          </div>
          <div
            class="space-y-2 flex flex-col"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
              for=":rb:-form-item"
            >
              Check-out
            </label>
            <button
              aria-controls="radix-:rc:"
              aria-describedby=":rb:-form-item-description"
              aria-expanded="false"
              aria-haspopup="dialog"
              aria-invalid="false"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[240px] pl-3 text-left font-normal text-muted-foreground"
              data-state="closed"
              disabled=""
              id=":rb:-form-item"
              type="button"
            >
              <span>
                Pick a date to your check-out
              </span>
              <svg
                class="lucide lucide-calendar ml-auto h-4 w-4 opacity-50"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  height="18"
                  rx="2"
                  ry="2"
                  width="18"
                  x="3"
                  y="4"
                />
                <line
                  x1="16"
                  x2="16"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="8"
                  x2="8"
                  y1="2"
                  y2="6"
                />
                <line
                  x1="3"
                  x2="21"
                  y1="10"
                  y2="10"
                />
              </svg>
            </button>
          </div>
          <div
            class="space-y-2 max-w-[240px]"
          >
            <label
              class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
              for=":rd:-form-item"
            >
              Guests
            </label>
            <input
              aria-describedby=":rd:-form-item-description"
              aria-invalid="false"
              class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
              id=":rd:-form-item"
              name="quantityGuests"
              placeholder="number of guests"
              type="number"
              value=""
            />
          </div>
          <div
            class="flex flex-col gap-4 mt-4"
          >
            <h4
              class="text-xl font-semibold tracking-tight"
            >
              Booking informations
            </h4>
            <div
              class="flex gap-6 justify-between"
            >
              <span
                class="underline"
              >
                $
                500
                 x 
                0
                 nights
              </span>
              <span>
                $
                0
              </span>
            </div>
            <div
              class="flex gap-6 justify-between"
            >
              <span
                class="underline"
              >
                $
                100
                 x 
                0
                 guests
              </span>
              <span>
                $
                0
              </span>
            </div>
            <div
              class="shrink-0 bg-border h-[1px] w-full max-w-[200px] mx-auto"
              data-orientation="horizontal"
              role="none"
            />
            <div
              class="flex gap-6 justify-between"
            >
              <span
                class="text-lg font-semibold"
              >
                Total
              </span>
              <span
                class="text-lg font-bold"
              >
                $
                0
              </span>
            </div>
          </div>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
            type="submit"
          >
            Book
          </button>
        </form>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 mt-4 w-full"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
